feat: introduce configurable `edit page on github` redirection

sambhavgupta0705 opened the pull request:
**Description**

- Some docs are imported from other repositories and currently if someone wants to fix some issues in it then they are redirected to website docs which are only a copy,
- for fixing the same a json file is setup for redirection to the correct url

**Related issue(s)**
<!-- If you refer to a particular issue, provide its number, otherwise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->
fixes #1175

sambhavgupta0705 committed:
feat: fixed the edit page on github redirection

sambhavgupta0705 committed:
general fixes

sambhavgupta0705 committed:
extra space removal

sambhavgupta0705 committed:
fixes

akshatnema commented on config/edit-page-config.json:
```diff
@@ -0,0 +1,13 @@
+
+
+]
+
```
```suggestion
[
    {
        "value": "/tools/generator",
        "href": "https://github.com/asyncapi/generator/tree/master/docs"
    },
    {
        "value": "reference/specification/",
        "href": "https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md"
    },
    {
        "value": "",
        "href": "https://github.com/asyncapi/website/blob/master/pages"
    }
]
```
This improves the readability of the JSON file.

akshatnema reviewed (changes requested):
Hey @sambhavgupta0705 thanks for raising the PR. I see certain concerns with this PR like when I click `Edit this page in github` on a certain doc of generator or specification, it doesn't take me to the specific file in the repo.  Can we improvise it in this manner?

Also mentioning @derberg for review and suggest the way to improve this.

sambhavgupta0705 committed:
Update config/edit-page-config.json

Co-authored-by: Akshat Nema <76521428+akshatnema@users.noreply.github.com>

sambhavgupta0705 commented:
Hi @derberg , I want to know the redirected files of Specs repo

derberg commented:
yeah we definitely should lint directly to file that someone wants to edit

> I want to know the redirected files of Specs repo
the one you have is good -> https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md
or I did not get the question 😄

sambhavgupta0705 commented:
> yeah we definitely should lint directly to file that someone wants to edit
> 
> > I want to know the redirected files of Specs repo
> > the one you have is good -> https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md
> > or I did not get the question smile

yeahh I want to know only that.
I was just in doubt that is there any specific files for different specs version.

derberg commented:
how about getting the file name from `post.id` ? it is always there. so take `id` and split by `/` and take last element of array

the best would be to simply extend `post` object in https://github.com/asyncapi/website/blob/master/scripts/build-post-list.js script, so you have all the data in `post` and later in `components/layout/DocsLayout.js` you have no magic

sambhavgupta0705 commented:
@derberg @akshatnema PTAL 
In the new commit I added a new variable last which will check the last endpoint after `/` and added a ternary operator for `specs` and `generator` doc

sambhavgupta0705 committed:
new solution with some automation

akshatnema reviewed (dismissed):
I approve this PR as I don't find anything where each specification version can be linked to the spec file. So, everything works fine for me in this PR. Also, review requested from @derberg.

derberg commented on components/layout/DocsLayout.js:
```diff
@@ -16,13 +16,23 @@ import AnnouncementHero from '../campaigns/AnnoucementHero'
+    return <a target="_blank" rel="noopener noreferrer" href={`${target.href}${post.isIndex ? post.slug + '/index' : post.slug}.md`} className="ml-1 underline">Edit this page on GitHub</a>
+
+  }
+  return <a target="_blank" rel="noopener noreferrer" href={`${target.href}/${target.value=="reference/specification/" ?'asyncapi.md':last}`} className="ml-1 underline">Edit this page on GitHub</a>
```
can we add logic that is not forcing us to do hardcode like `${target.value=="reference/specification/" ?'asyncapi.md':last}` ?

so support:
```
{
        "value": "reference/specification/",
        "href": "https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md"
    }
```

and you check, if the last element of the `href` is a filename, then `Edit the page on GitHub` always point to it on pages from `reference/specification/"`

sambhavgupta0705 replied:
I didn't find any other way then hard coding because we there is only a single file for `reference/specification` and many other directions for other file .

One way to remove ternary operator is to have an if statement for file.

It would be nice if you could suggest some ways

derberg reviewed (commented):
one small improvement possible, but important imho

sambhavgupta0705 committed:
Merge branch 'master' into editpageFix

akshatnema commented:
ping @sambhavgupta0705

sambhavgupta0705 committed:
Merge branch 'master' into editpageFix

sambhavgupta0705 committed:
Merge branch 'master' into editpageFix

sambhavgupta0705 commented:
@derberg PTAL
- added a check for `reference/specification` (need your suggestions for that)
- added redirection for cli docs

sambhavgupta0705 committed:
fix: reference/specification href fix

sambhavgupta0705 committed:
Revert "fix: reference/specification href fix"

This reverts commit 045244d5600eb7211d46b9e621c11108afa3bf91.

sambhavgupta0705 committed:
fix: fixed the reference/specification link

derberg commented on components/layout/DocsLayout.js:
```diff
@@ -17,13 +17,24 @@ import { SearchButton, DOCS_INDEX_NAME } from '../AlgoliaSearch';
+  }
+  if (target.value=="reference/specification/") last=""
+  
+  return <a target="_blank" rel="noopener noreferrer" href={`${target.href}/${last}`} className="ml-1 underline">Edit this page on GitHub</a>
```
`Edit this page on GitHub` - duplication, better put this label in a variable

derberg commented on components/layout/DocsLayout.js:
```diff
@@ -17,13 +17,24 @@ import { SearchButton, DOCS_INDEX_NAME } from '../AlgoliaSearch';
-  return <a target="_blank" rel="noopener noreferrer" href={`https://github.com/asyncapi/website/blob/master/pages${post.isIndex ? post.slug + '/index' : post.slug}.md`} className="ml-1 underline">Edit this page on GitHub</a>
-}
+
+  var last=post.id.substring(post.id.lastIndexOf("/") + 1);
```
`var` 👀

derberg commented on components/layout/DocsLayout.js:
```diff
@@ -17,13 +17,24 @@ import { SearchButton, DOCS_INDEX_NAME } from '../AlgoliaSearch';
+    return <a target="_blank" rel="noopener noreferrer" href={`${target.href}${post.isIndex ? post.slug + '/index' : post.slug}.md`} className="ml-1 underline">Edit this page on GitHub</a>
+
+  }
+  if (target.value=="reference/specification/") last=""
```
`reference/specification/` we still have hardcode in code, so config is not totally independent.

you can do something like this
pseudo code incoming:
```
const editHref = target.href;
//in case href points to a file, then edit link should always point to that file and not be dynamic
const hrefList = editHref.split('/');
const lastListElement = hrefList[hrefList.length - 1].split('.');
const isHrefToFile = lastListElement.length > 1;
```

so instead of 
```
if (target.value=="reference/specification/") last=""
```
you could do
```
if (isHrefToFile) last = ""
```

sambhavgupta0705 committed:
removed hardcoded changes

sambhavgupta0705 committed:
Merge branch 'master' into editpageFix

sambhavgupta0705 commented:
@derberg it works fine now
PTAL

derberg reviewed (approved):
lgtm 👏🏼

derberg committed:
Merge branch 'master' into editpageFix

derberg commented:
/rtm
